File: src/main/java/de/fluxparticle/animation/graph/GraphLevelSolver.java
package de.fluxparticle.animation.graph;

import java.util.Arrays;
import java.util.Comparator;

import static java.lang.Integer.compare;

public class GraphLevelSolver {

	private static class Range {

		private int min;

		private int max;

		public Range(int min, int max) {
			super();
			this.min = min;
			this.max = max;
		}

		public int getMin() {
			return min;
		}

		public void setMin(int min) {
			if (min > this.min) {
				this.min = min;
			}
		}

		public int getMax() {
			return max;
		}

		public void setMax(int max) {
			if (max < this.max) {
				this.max = max;
			}
		}

		public boolean isEmpty() {
			return min > max;
		}

		public int getCenter() {
			return (min + max) / 2;
		}

	}

	private static final Comparator<GraphNode> COMPARATOR = (o1, o2) -> compare(o1.getPreferredPos(), o2.getPreferredPos());

	private final GraphNode[] nodes;

	private final int maxK;

	private final int min;

	private final int max;

	public GraphLevelSolver(GraphNode[] nodes) {
		this.nodes = nodes;

		Arrays.sort(nodes, COMPARATOR);

		int maxK = 0;
		for (GraphNode graphNode : nodes) {
			maxK += graphNode.getSize();
		}
		this.maxK = maxK;

		min = nodes[0].getPreferredPos() - maxK;
		max = nodes[nodes.length-1].getPreferredPos() + maxK;
	}

	public void solve() {
		int k = findK();

		Range[] ranges = mkRanges(k);

		for (int i = 0; i < ranges.length; i++) {
			nodes[i].setPos(ranges[i].getCenter());
		}
	}

	private int findK() {
		int lo = 0;
		int hi = maxK;

		while (lo < hi) {
			int k = (lo + hi) / 2;
			if (check(k)) {
				hi = k;
			} else {
				lo = k + 1;
			}
		}

		return hi;
	}

	private boolean check(int k) {
		Range[] ranges = mkRanges(k);

		for (int i = 0; i < ranges.length; i++) {
			if (ranges[i].isEmpty()) {
				return false;
			}
		}

		return true;
	}

	private Range[] mkRanges(int k) {
		Range[] ranges = new Range[nodes.length];

		for (int i = 0; i < ranges.length; i++) {
			ranges[i] = new Range(min, max);
		}

		for (int i = 0; i < ranges.length; i++) {
			ranges[i].setMin(nodes[i].getPreferredPos() - k);
			ranges[i].setMax(nodes[i].getPreferredPos() + k);
		}

		for (int i = 1; i < ranges.length; i++) {
			ranges[i].setMin(ranges[i-1].getMin() + nodes[i-1].getSize());
		}

		for (int i = ranges.length - 2; i >= 0; i--) {
			ranges[i].setMax(ranges[i+1].getMax() - nodes[i].getSize());
		}

		return ranges;
	}

}

File: src/main/kotlin/de/fluxparticle/animation/graph/GraphLevelSolver.kt
package de.fluxparticle.animation.graph

import java.lang.Integer.compare
import java.util.*

class GraphLevelSolver(private val nodes: Array<GraphNode>) {

    private val maxK: Int

    private val min: Int

    private val max: Int

    private class Range(private var min: Int, private var max: Int) {

        val isEmpty: Boolean
            get() = min > max

        val center: Int
            get() = (min + max) / 2

        fun getMin(): Int {
            return min
        }

        fun setMin(min: Int) {
            if (min > this.min) {
                this.min = min
            }
        }

        fun getMax(): Int {
            return max
        }

        fun setMax(max: Int) {
            if (max < this.max) {
                this.max = max
            }
        }

    }

    init {
        Arrays.sort(nodes, COMPARATOR)

        var maxK = 0
        for (graphNode in nodes) {
            maxK += graphNode.size
        }
        this.maxK = maxK

        min = nodes[0].preferredPos - maxK
        max = nodes[nodes.size - 1].preferredPos + maxK
    }

    fun solve() {
        val k = findK()

        val ranges = mkRanges(k)

        for (i in ranges.indices) {
            nodes[i].setPos(ranges[i].center)
        }
    }

    private fun findK(): Int {
        var lo = 0
        var hi = maxK

        while (lo < hi) {
            val k = (lo + hi) / 2
            if (check(k)) {
                hi = k
            } else {
                lo = k + 1
            }
        }

        return hi
    }

    private fun check(k: Int): Boolean {
        val ranges = mkRanges(k)

        for (i in ranges.indices) {
            if (ranges[i].isEmpty) {
                return false
            }
        }

        return true
    }

    private fun mkRanges(k: Int): Array<Range> {
        val ranges = Array(nodes.size) {
            Range(min, max)
        }

        for (i in ranges.indices) {
            ranges[i].setMin(nodes[i].preferredPos - k)
            ranges[i].setMax(nodes[i].preferredPos + k)
        }

        for (i in 1 until ranges.size) {
            ranges[i].setMin(ranges[i - 1].getMin() + nodes[i - 1].size)
        }

        for (i in ranges.size - 2 downTo 0) {
            ranges[i].setMax(ranges[i + 1].getMax() - nodes[i].size)
        }

        return ranges
    }

    companion object {

        private val COMPARATOR = { o1: GraphNode, o2: GraphNode -> compare(o1.preferredPos, o2.preferredPos) }

    }

}
